xc/xenops_server_xen.ml

```diff
@@ -2339,6 +2370,66 @@ module VIF = struct
 				if di.Xenctrl.hvm_guest
 				then ignore (run !Xc_path.setup_vif_rules ["classic"; tap_interface_name; vm; devid; "filter"])
 			)
+	
+	let set_ipv4_configuration task vm vif ipv4_configuration =
+		let open Device_common in
+		with_xc_and_xs
+			(fun xc xs ->
+				let device = device_by_id xc xs vm Vif Newest (id_of vif) in
+				let domid = device.frontend.domid in
+				let devid = string_of_int device.frontend.devid in
+				let xenstore_path = Printf.sprintf "/local/domain/%d/xenserver/device/vif/%s/static-ip-setting" domid devid in
+				match ipv4_configuration with
+				| Vif.Unspecified4 ->
+					let ip_setting_enabled = Printf.sprintf "%s/%s" xenstore_path "enabled" in
+					debug "xenstore-write %s <- %s" ip_setting_enabled "0";
+					xs.Xs.write ip_setting_enabled "0"
+				| Vif.Static4 (address, gateway) ->
+					let ip_setting_enabled = Printf.sprintf "%s/%s" xenstore_path "enabled" in
+					debug "xenstore-write %s <- %s" ip_setting_enabled "1";
+					xs.Xs.write ip_setting_enabled "1";
```

Comment:
To avoid races, all these keys should be written in a single xenstore transaction. See e.g. `Generic.device_add` in device.ml.